### PR TITLE
[cmusphinx][WIP] CMU Sphinx Speech-to-Text initial contribution

### PR DESCRIPTION
**Pull request by ghys**
This is a standalone, pure Java implementation of a `STTService` using [CMU Sphinx](http://cmusphinx.sourceforge.net) based on the existing one for Kaldi. I didn't choose pocketsphinx for simplicity. It is left disabled in the parent POM as well since it's probably not working. I did it out of pure curiosity and since I wasn't aware of any public effort to build one, but I'm unsure how to use it in practice, so it is probably incomplete. Still, you can run the unit tests in `CMUSphinxSTTServiceTest.java` if you're interested. Feel free to close this PR if it isn't relevant.

Here's how your can run the unit tests:
- You need to prepare the following structure on your filesystem:

```
./de
./de/cmusphinx-de-ptm-voxforge-5.2
./de/cmusphinx-de-ptm-voxforge-5.2/AUTHORS
./de/cmusphinx-de-ptm-voxforge-5.2/COPYING
./de/cmusphinx-de-ptm-voxforge-5.2/feat.params
./de/cmusphinx-de-ptm-voxforge-5.2/mdef
./de/cmusphinx-de-ptm-voxforge-5.2/means
./de/cmusphinx-de-ptm-voxforge-5.2/mixture_weights
./de/cmusphinx-de-ptm-voxforge-5.2/noisedict
./de/cmusphinx-de-ptm-voxforge-5.2/README.md
./de/cmusphinx-de-ptm-voxforge-5.2/sendump
./de/cmusphinx-de-ptm-voxforge-5.2/transition_matrices
./de/cmusphinx-de-ptm-voxforge-5.2/variances
./de/cmusphinx-voxforge-de.dic
./de/grammar
./de/grammar/digits_months.gram
./en
./en/cmudict-en-us.dict
./en/cmusphinx-en-us-ptm-5.2
./en/cmusphinx-en-us-ptm-5.2/feat.params
./en/cmusphinx-en-us-ptm-5.2/mdef
./en/cmusphinx-en-us-ptm-5.2/means
./en/cmusphinx-en-us-ptm-5.2/mixture_weights
./en/cmusphinx-en-us-ptm-5.2/noisedict
./en/cmusphinx-en-us-ptm-5.2/README
./en/cmusphinx-en-us-ptm-5.2/sendump
./en/cmusphinx-en-us-ptm-5.2/transition_matrices
./en/cmusphinx-en-us-ptm-5.2/variances
./en/en-70k-0.2-pruned.lm
./fr
./fr/cmusphinx-fr-ptm-5.2
./fr/cmusphinx-fr-ptm-5.2/feat.params
./fr/cmusphinx-fr-ptm-5.2/mdef
./fr/cmusphinx-fr-ptm-5.2/means
./fr/cmusphinx-fr-ptm-5.2/mixture_weights
./fr/cmusphinx-fr-ptm-5.2/noisedict
./fr/cmusphinx-fr-ptm-5.2/README
./fr/cmusphinx-fr-ptm-5.2/sendump
./fr/cmusphinx-fr-ptm-5.2/transition_matrices
./fr/cmusphinx-fr-ptm-5.2/variances
./fr/fr.dict
./fr/french-small.lm
./fr/grammar
./fr/grammar/digits_months.gram
```

using:
1. acoustic models, dictionaries, and language models (optional for French and German since we use grammar files instead) downloaded from https://sourceforge.net/projects/cmusphinx/files/Acoustic%20and%20Language%20Models/ (refer to the file names to figure out which ones to download);
2. custom grammar files for French and German (currently the service is unable to use the grammar passed as a parameter to the `recognize()` method) like these ones:

_fr/grammar/digits_months.gram_:
```
#JSGF V1.0;

grammar digits_months;

<digit> = un | deux | trois | quatre | cinq | six | sept | huit | neuf | dix;
<month> = janvier | février | mars | avril | mai | juin | juillet | août | septembre | octobre | novembre | décembre;

public <digit_month> = <digit> | <month>;
```
_de/grammar/digits_months.gram_:
```
#JSGF V1.0;

grammar digits_months;

<digit> = eins | zwei | drei | vier | fünf | sechs | sieben | acht | neun | zehn;
<month> = januar | februar | märz | april | mai | juni | juli | august | september | oktober | november | dezember;

public <digit_month> = <digit> | <month>;
```

* Change the `resourcesBasePath` in `CMUSphinxSTTServiceTest.java` line 25 to the location you put the files above.

Add jUnit to the classpath and run the tests, it will use the sample files from https://www.freesound.org in the test project's resources.

The test using the language model is really too slow and resource-intensive, but the ones with the limited grammar (which could make sense for home automation commands) run a lot faster. The accuracy is not great but just OK.

This pastebin has the output of a successful test run to evaluate the recognition results (I wasn't able to get the confidence score right so it's hardcoded to 1...): http://pastebin.ubuntu.com/24479582/

Signed-off-by: Yannick Schaus <habpanel@schaus.net>

### CODE REVIEW

**Comment by JochenHiller** [comment]
@ghys  Looks like interesting stuff.
I just quick viewed the PR: Is it really a pure Java implementation? At least I could not find any use of native libraries. That would be cool.
On which hardware platform did you use that at least for playing around? Is a RaspPi3 similar environment feasible, or is the quality strong dependent on hardware capabilities?

**Comment by ghys** [comment]
@JochenHiller yes it is a pure Java implementation - as opposed to pocketsphinx which is a native library but apparently offers better performance.

I found that performance and accuracy depend a lot on the use of a grammar file, if you don't provide one, it struggles even on my desktop PC and uses a lot of CPU and RAM:
```
confidence: 1,000000, transcript: this is my voice that is speaking to you
INFO speedTracker            Total Time Audio: 2,38s  Proc: 7,97s 3,35 X real time
INFO memoryTracker           Used: This: 878,68 Mb  Avg: 875,57 Mb  Max: 878,68 Mb
```

However, if you do ([examples](http://cmusphinx.sourceforge.net/wiki/sphinx4:jsgfsupport)) it will be a lot better - but you have to explicitely tell the phrases and words you want it to recognize. I think we could come up with one covering most of the use cases.
```
org.eclipse.smarthome.core.voice.SpeechRecognitionEvent
confidence: 1,000000, transcript: eins
org.eclipse.smarthome.core.voice.SpeechRecognitionEvent
confidence: 1,000000, transcript: zwei
org.eclipse.smarthome.core.voice.SpeechRecognitionEvent
confidence: 1,000000, transcript: drei
org.eclipse.smarthome.core.voice.SpeechRecognitionEvent
confidence: 1,000000, transcript: vier
org.eclipse.smarthome.core.voice.SpeechRecognitionEvent
confidence: 1,000000, transcript: fünf
org.eclipse.smarthome.core.voice.SpeechRecognitionEvent
confidence: 1,000000, transcript: sechs
org.eclipse.smarthome.core.voice.SpeechRecognitionEvent
confidence: 1,000000, transcript: sieben
org.eclipse.smarthome.core.voice.SpeechRecognitionEvent
confidence: 1,000000, transcript: acht
org.eclipse.smarthome.core.voice.SpeechRecognitionEvent
confidence: 1,000000, transcript: neun
org.eclipse.smarthome.core.voice.SpeechRecognitionEvent
confidence: 1,000000, transcript: zehn
INFO speedTracker            Total Time Audio: 7,05s  Proc: 0,99s 0,14 X real time
INFO memoryTracker           Used: This: 130,61 Mb  Avg: 105,07 Mb  Max: 166,05 Mb
```

I have yet to try it on my RPi, but I fear it's going to depend a lot on the size of the acoustic model, dictionary and grammar you feed it.

**Comment by kaikreuzer** [comment]
Hi @ghys,

This is pretty interesting!

> using CMU Sphinx based on the existing one for Kaldi.

Note that there is also a PocketSphinx implementation in here: https://github.com/mozilla/smarthome/tree/master/bundles/io/org.eclipse.smarthome.io.voice
@kdavis-mozilla had this working with openHAB, but I struggled getting the C code of pocketsphinx compiled and working smoothly.

> I didn't choose pocketsphinx for simplicity.

Yes, a pure Java implementation is indeed much more desirable for openHAB. I understood from @kdavis-mozilla that the accuracy and performance of the C version is much better, though.
Nonetheless, if we manage to get a decent performance out of the Java version for a limited vocabulary, this would be very valuable indeed. It probably makes sense to sync the grammar with our [StandardInterpreter](https://github.com/eclipse/smarthome/blob/master/bundles/core/org.eclipse.smarthome.core.voice/src/main/java/org/eclipse/smarthome/core/voice/internal/text/StandardInterpreter.java).

I didn't quite get in what state your PR is: Is it already working in some way? What would we need to add to work with a given vocabulary? I think it would be awesome to have that available in openHAB, as an offline STT is one of the lacking features.

@kdavis-mozilla Even if you are not working on that stuff anymore, maybe you find a few minutes of your spare time to chime in here :-)

**Comment by ghys** [comment]
> It probably makes sense to sync the grammar with our StandardInterpreter.

Agreed!

> I didn't quite get in what state your PR is: Is it already working in some way? What would we need to add to work with a given vocabulary?

Well it will supposedly send STTEvents to a provided `STTListener` but I had trouble figuring out how it all works, namely how (in practice) to capture audio from an audio source and pass it to the STT service then to the HLI. I realize I also missed the whole keyword spotting stuff, though I shouldn't be too hard to do. Using a grammar provided to the `recognize()` method instead of one on the filesystem should also be possible.

What exactly are the AudioStreams passed to the STT Service? Little bits of audio or a continuous stream from the source?

>  I think it would be awesome to have that available in openHAB, as an offline STT is one of the lacking features.

This would be my goal too - voice control (even a basic one) easy to deploy and not depending on a cloud service!

**Comment by ghys** [comment]
I have added keyword spotting and the ability to start `VoiceManager.startDialog()` automatically when the bundle is activated (let me know if there's a better way to do this, I haven't found one).

I'm quite pleased by the results, it works pretty well, I tested it - in French - on my development machine with the following configuration:
```
org.eclipse.smarthome.audio:defaultSink=javasound
org.eclipse.smarthome.audio:defaultSource=javasound
org.eclipse.smarthome.core.localeprovider:language=fr
org.eclipse.smarthome.core.localeprovider:region=FR
org.eclipse.smarthome.voice:defaultHLI=system
org.eclipse.smarthome.voice:defaultSTT=cmusphinx
org.eclipse.smarthome.voice:defaultTTS=voicerss
org.eclipse.smarthome.voice:defaultVoice=voicerss:frFR
org.eclipse.smarthome.voice:defaultKS=cmusphinx
org.eclipse.smarthome.voice:keyword=maison

# - download acoustic model and dictionary from
# https://sourceforge.net/projects/cmusphinx/files/Acoustic%20and%20Language%20Models/
# - create the grammar path (folder)
# - create a commands.gram file in the grammar folder
# - change paths below
# - uncomment the last line to start dialogging when the bundle is activated
org.openhab.cmusphinxsst:locale=fr-FR
org.openhab.cmusphinxsst:acousticModelPath=I:/dev/stt/fr/cmusphinx-fr-ptm-5.2
org.openhab.cmusphinxsst:dictionaryPath=I:/dev/stt/fr/fr.dict
org.openhab.cmusphinxsst:grammarPath=I:/dev/stt/fr/grammar/
org.openhab.cmusphinxsst:grammarName=commands
#org.openhab.cmusphinxsst:startDialog=true
```
with this grammar matching part of the French version by @lolodomo in `StandardInterpreter` (and including the names of a few of my items):
```
#JSGF V1.0;

grammar commands;

<command> = allumer | éteindre | activer | éteindre | stopper | désactiver | couper | augmenter | diminuer | monter | descendre;
<lela> = le | la | les | l;
<poursurde> = pour | sur | du | de;

<color> = blanc | rose | jaune | orange | violet | rouge | vert | bleu;

<item> = bureau | salon | table | chambre | cuisine | volet;

<keyword> = maison;

public <order> = <keyword> | <command> [<lela>] <item> | couleur <color> [<poursurde>] [<lela>] <item>;
```
I can now say things like "_maison, éteindre le bureau_" or "_maison, couleur orange pour salon_" into the mic and it will execute the command - which is pretty cool 😉 

I also filed issues https://github.com/eclipse/smarthome/issues/3347 and https://github.com/eclipse/smarthome/issues/3348 in ESH as I ran into them while I was testing all this.

**Comment by ghys** [comment]
I'd be interested in results from other beta testers :)

I added a [README](https://github.com/ghys/openhab2-addons/blob/cmusphinx/addons/voice/org.openhab.voice.cmusphinx/README.md) file to help you in configuring this as it stands now.
The grammar still has to be supplied by the user for the time being, since there doesn't seem to be a way to pass the grammar to the `recognize()` method currently.

Let me know if I can be of any help.

**Comment by kaikreuzer** [comment]
> I have added keyword spotting and the ability to start VoiceManager.startDialog() automatically when the bundle is activated (let me know if there's a better way to do this, I haven't found one).

I think the idea here was to leave it to the user/setup when the keyword spotting should start. Also the startDialog can be called with parameters, if non-default services should be used.
So imho it could make sense to add the call to VoiceConsoleCommandExtension and as a new module handler in MediaModuleHandlerFactory. The KS service itself imho should not have any dependency on the VoiceManager. We could rather have a config parameter for the VoiceManager that allows to always start a dialog with the default services.

> I'd be interested in results from other beta testers :)

I'll try to test it soon as I am VERY interested in this feature as well! Many thanks for pushing this forward!

> The grammar still has to be supplied by the user for the time being, since there doesn't seem to be a way to pass the grammar to the recognize() method currently.

I agree, we need a way to feed this in (and to share it with the HLI service).

**Comment by ghys** [comment]
> I think the idea here was to leave it to the user/setup when the keyword spotting should start.

I understand controlling the VoiceManager from the bundle it a dirty solution, but it was intended it to be only temporary during the development/alpha testing phase; I was simply eager to have it working quickly :), and tried to avoid changes in ESH for now, at least until a working approach and interfaces are agreed upon. The discussion should take place in an ESH issue anyway.

I'll remove this part and update the documentation when there's a "good" way of starting the keyword spotting.

> Also the startDialog can be called with parameters, if non-default services should be used.

Yes - I only had a problem with the keyword ([a null value is not replaced by the default/configured one](https://github.com/eclipse/smarthome/blob/3ae5d7b1b284f2bde6199bfe8dad950fd38cae39/bundles/core/org.eclipse.smarthome.core.voice/src/main/java/org/eclipse/smarthome/core/voice/internal/VoiceManagerImpl.java#L370), like the other parameters). I'll submit a PR for this ;)

> I agree, we need a way to feed this in (and to share it with the HLI service).

Yes. The high level API only supports loading grammar from files ([Javadoc](http://cmusphinx.sourceforge.net/doc/sphinx4/edu/cmu/sphinx/api/Configuration.html)), although it's apparently possible to [load and switch grammars dynamically](http://cmusphinx.sourceforge.net/wiki/sphinx4:swappinggrammars).
Also JSGF supports imports: https://www.w3.org/TR/jsgf/#17971. This could be interesting for dynamically generated parts like item names, imported into a more static grammar. This would also allow the user to define their own local (JSGF) rules, useful for instance if the rule-based HLI is used (e.g. for specific sentences like "watch a movie" or "I'm going away"). In any case, the STT and the HLI functionally must agree on the same grammar to work together.

> I'll try to test it soon as I am VERY interested in this feature as well! Many thanks for pushing this forward!

Pleasure. If you only want to try it out and can't be bothered to clone and build my fork, which I completely understand 😄:
[Here's a binary](https://github.com/ghys/openhab2-addons/releases/tag/2.1.0-20170503-cmusphinx)

**Comment by ghys** [comment]
Just a quick update: I bought a cheap USB microphone at Saturn (http://www.saturn.de/de/product/_trust-20378-2031514.html) to hook up to a Raspberry Pi clone (Cortex-A7 dual-core, 1GB RAM) in the living room, see how the STT performs on these kind of devices. The results are encouraging and much better than I expected! It takes a good 10-15 seconds to initialize the bundle, but after that it mostly works, doesn't lag and doesn't seem to hog resources. Of course, the mic quality is paramount, for 20€ I got what I paid for, but if I'm rather close (<3-4m) to the mic, no background noise, limit the grammar to a few distinguishable sentences, it will pick up the keyword reliably and most of what I say. 🎉

**Comment by kaikreuzer** [comment]
That really sounds encouraging and much better than things I have heard from others about the to be expected quality. I hope I'll find time tomorrow to test it as well!

**Comment by kaikreuzer** [comment]
> Pleasure. If you only want to try it out and can't be bothered to clone and build my fork, which I completely understand 😄:
> Here's a binary

Good start - do you maybe also have an example cmusphinx.cfg file and contents of acousticModelPath and dictionaryPath?

**Comment by ghys** [comment]
I actually used Paper UI to configure it, like so:

image

The equivalent `conf/services/cmusphinx.cfg` file would be (if I'm not mistaken):
```
locale=fr-FR
acousticModelPath=/opt/openhab/stt/fr/cmusphinx-fr-ptm-5.2
dictionaryPath=/opt/openhab/stt/fr/fr.dict
grammarPath=/opt/openhab/stt/fr/grammar/
grammarName=commands
startDialog=true
```

For German you would use the extracted contents of *cmusphinx-de-ptm-voxforge-5.2.tar.gz* as the acoustic model and *cmusphinx-voxforge-de.dic* as the dictionary - to download from https://sourceforge.net/projects/cmusphinx/files/Acoustic%20and%20Language%20Models/German/ (ignore the readme there).

Lastly, default audio and voice services defaults have to be set (before switching startDialog to true) as well as a matching localeProvider:

image
image
image

**Comment by kubawolanin** [comment]
Awesome stuff Yannick! Already downloaded your binding, but since there's no polish acoustic model for CMU Sphinx, I started making one ;-) You've inspired me.
So it'll take some time before I jump on the board. I hope I'll have some decent results - I have a 13h audiobook already transcribed and around 40 mins of sliced sentences, extracted from each chapter 😨 

Keep up the great work!

**Comment by ghys** [comment]
Wow that's a massive undertaking @kubawolanin - good luck! :)
Let me know how it works for you!

In other news, I temporarily disabled it from my "production" install - too many false positives, resulting in random stuff sometimes happening like when watching TV ;)

What happens is, it's spotting the keyword (wrongfully), then interprets the next hypothesis if it's valid, even if it's several minutes later. There should be some kind of timeout (10-15s) if nothing is heard after spotting the keyword.

**Comment by kaikreuzer** [comment]
@ghys Thanks for your guidance, I got it working!

The two ESH PRs from you are by now available in the openHAB distro (which makes your binary above not 100% compatible anymore because of the change in the KS interface).

> although it's apparently possible to load and switch grammars dynamically.

This is definitely what we would need next. So in order to sync HLI and SST, we have to decide for a grammar format (or rather define a Java Object Model for it). JSGF might be one choice (possibly with the [Sphinx object model](http://cmusphinx.sourceforge.net/doc/sphinx4/edu/cmu/sphinx/jsgf/JSGFGrammar.html)), [grXML](https://www.w3.org/TR/speech-grammar/) seems to be another one (maybe more widespread, because not tied to Java?), which also seems to be [supported by Sphinx](http://cmusphinx.sourceforge.net/doc/sphinx4/edu/cmu/sphinx/jsgf/GrXMLGrammar.html). What do you think?

When this is defined, we should also create an extension mechanism in the Sphinx bundle that allows us to register acoustic models and dictionaries dynamically through additional bundles that can provide such resources - that would make the setup much easier for users.

> There should be some kind of timeout (10-15s) if nothing is heard after spotting the keyword.

Yes, definitely - can you come up with a PR for ESH for this?

**Comment by ghys** [comment]
> > There should be some kind of timeout (10-15s) if nothing is heard after spotting the keyword.

> Yes, definitely - can you come up with a PR for ESH for this?

Sorry I missed your message and did the changes in the addon itself (commit above). Actually IMHO it should be the responsibility of the STT Service itself to handle the timeout (since it could be implementation-specific), and the current implementation of the `DialogProcessor` is ESH is fine: it simply [awaits for a `RecognitionStopEvent`](https://github.com/eclipse/smarthome/blob/master/bundles/core/org.eclipse.smarthome.core.voice/src/main/java/org/eclipse/smarthome/core/voice/internal/DialogProcessor.java#L131) to go back to the keyword spotting state, so it works as long as this event is eventually sent (which could be expected/a requirement of all STT Services implementations).

> The two ESH PRs from you are by now available in the openHAB distro (which makes your binary above not 100% compatible anymore because of the change in the KS interface).

Yes, the above commit also includes this ([binary here](https://github.com/ghys/openhab2-addons/releases/tag/2.1.0-20170509-cmusphinx)).

> JSGF might be one choice (possibly with the Sphinx object model), grXML seems to be another one (maybe more widespread, because not tied to Java?), which also seems to be supported by Sphinx.

I agree with your argument about grXML - but I must admit I have a personal preference for BNF-style grammars i.e. JSGF, for readability and verbosity reasons...
https://www.w3.org/TR/speech-grammar/ defines an ABNF alternative style similar to JSGF but it doesn't seem to be supported by Sphinx.

I also found a class supposed to generate a JSGF grammar from HLI rules while browsing ESH code:
https://github.com/eclipse/smarthome/blob/master/bundles/core/org.eclipse.smarthome.core.voice/src/main/java/org/eclipse/smarthome/core/voice/text/AbstractRuleBasedInterpreter.java#L619.
Do you know if it is working/reliable? It we can generate a JSGF file for instance in `userdata` we could simply load it with Sphinx's high level API and avoid delving into the grammar APIs?

> When this is defined, we should also create an extension mechanism in the Sphinx bundle that allows us to register acoustic models and dictionaries dynamically through additional bundles that can provide such resources - that would make the setup much easier for users.

Good idea! I thought about this too but beware of the size (>10MB) and their licensing - haven't looked much at the licenses but they seem to vary from CMU Sphinx's.

**Comment by kubawolanin** [comment]
@ghys the diff of this PR became really big. Perhaps you've merged master along with 81e6188294cc80a339627d01d4653fea09be4b92 ?

**Comment by ghys** [comment]
Yikes, I messed up my rebase 😕

**Comment by kaikreuzer** [comment]
> Actually IMHO it should be the responsibility of the STT Service itself to handle the timeout

Yes, you are right.

> I agree with your argument about grXML - but I must admit I have a personal preference for BNF-style grammars

I agree with that - when manually writing files, I also prefer BNF, but my question was more about the data model for the API itself.

> Do you know if it is working/reliable?

No idea, I wasn't even aware of its existence :-)

>  It we can generate a JSGF file for instance in userdata we could simply load it with Sphinx's high level API and avoid delving into the grammar APIs?

I would prefer working with a Java object model that we can hand to the different services; if we rely on code generation, this would have to be kept in sync with changes of items, rules, etc, which might be more complicated in the long run.

>  I thought about this too but beware of the size (>10MB) and their licensing

Sounds like perfect candidates for the Eclipse IoT Marketplace then 😎

**Comment by martinvw** [comment]
@ghys is this something which should be reviewed or did you just want to share it. If its the latest then it might be more explicit if we add [WIP] at the beginning of the title so people know what the status is.

**Comment by ghys** [comment]
@martinvw you're right, its status is somewhat confusing.
It started as an experiment and I filed the PR mainly to create a space here for discussion. Besides, I haven't been able to work on it for the last couple of weeks. I added the [WIP] prefix as you suggested.

To summarise, it's *mostly* working currently (*) but expects the user to provide the needed additional resources (acoustic model, dictionary, grammar); @kaikreuzer argued it would better to have an user-friendlier way of provisioning those - additional bundles for the acoustic model & dictionary, dynamic generation from the HLI for the grammar. Those are good ideas and perhaps should be implemented before talking about reviews and merging. I'll try to find some time next week to work on it.

(*) I should add a grain of salt here: it works well in small rooms (up to 15m²) with a conventional microphone, or you can buy a [PlayStation Eye](https://en.wikipedia.org/wiki/PlayStation_Eye) for under 10€ (a popular option for DIY voice assistant apparently since it has a decent 4-microphone array for the price).
In larger rooms, it's having issues with accuracy unless you're close to the microphone; worse it even sometimes picks up the keyword from background noise. Still experimenting to improve this... 
In short, don't get your hopes up too soon - this is unsurprisingly no Alexa/Echo killer especially without a decent far-field microphone array. Maybe [this Kickstarter project](https://www.seeedstudio.com/ReSpeaker-Mic-Array-Far-field-w%2F-7-PDM-Microphones-p-2719.html) is an option...

**Comment by martinvw** [comment]
@ghys Please note that, because it's marked [WIP], I do and did not review this PR completely yet.

If it's finished and ready to review please update the title and mention me so I can remove the "Work in progress" label.

**Comment by ghys** [comment]
Noted. I'll leave the [WIP] for now, I'm waiting for new hardware (far-field microphone) to conduct further real-world testing.

**Comment by hoalex** [comment]
Any new progress on this? Having offline STT in openHAB would be really nice, as to avoid using commercial providers like Apple/Amazon/Google. 👍

**Comment by martinvw** [comment]
@ghys do you plan to wrap this up, or is just open for any volunteer to take over?

**Comment by ghys** [comment]
To be perfectly honest, I don't have an use for it right now, and configuring it is hard (which would mean more support would be required), so it's pretty much abandoned until further notice... But I have some long term plans for it in conjunction with HABot, I just don't know when I'll get around to them.
If anyone is interested and willing to take over I'd gladly pass the torch :)

**Comment by Hilbrand** [comment]
I propose to close this pr for now as there doesn't seem to be an estimated time when it's finished and it probably require some work to get everything up-to-date. The code will still be available. Therefore if someone wants to progress one can still use the code from here.
